frontend/src/BudgetCard.tsx

```diff
@@ -0,0 +1,225 @@
+import * as React from "react";
+import Card from "@mui/joy/Card";
+import Checkbox from "@mui/joy/Checkbox";
+import Grid from "@mui/joy/Grid";
+import Typography from "@mui/joy/Typography";
+import { BudgetInput } from "./BudgetInput";
+import {
+    BudgetFilter,
+    getFighterCount,
+    getFleetSupplyRemaining,
+    getTotalCost,
+    sumUnitCounts,
+    UnitCounts,
+} from "./data";
+import { OverUnderNumber } from "./OverUnderNumber";
+import { GridCell } from "./GridCell";
+import { BudgetRow } from "./BudgetRow";
+
+interface Props {
+    unitCounts: UnitCounts;
+    budgetFilters: Set<BudgetFilter>;
+    resourceBudget: number;
+    setResourceBudget: (value: number) => void;
+    capacityBudget: number;
+    setCapacityBudget: (value: number) => void;
+    currentFleetSupply: number;
+    setCurrentFleetSupply: (value: number) => void;
+    maxFleetSupply: number;
+    setMaxFleetSupply: (value: number) => void;
+    isFighterUpgraded: boolean;
+    setIsFighterUpgraded: (value: boolean) => void;
+    shipCapacityUsed: number;
+    setShipCapacityUsed: (value: number) => void;
+    maxShipCapacity: number;
+    setMaxShipCapacity: (value: number) => void;
+    spaceDockFighterBonus: number;
+    setSpaceDockFighterBonus: (value: number) => void;
+}
+
+export function BudgetCard({
+    unitCounts,
+    budgetFilters,
+    resourceBudget,
+    setResourceBudget,
+    capacityBudget,
+    setCapacityBudget,
+    currentFleetSupply,
+    setCurrentFleetSupply,
+    maxFleetSupply,
+    setMaxFleetSupply,
+    isFighterUpgraded,
+    setIsFighterUpgraded,
+    shipCapacityUsed,
+    setShipCapacityUsed,
+    maxShipCapacity,
+    setMaxShipCapacity,
+    spaceDockFighterBonus,
+    setSpaceDockFighterBonus,
+}: Props) {
+    const fighterCapacityRemaining =
+        maxShipCapacity +
+        spaceDockFighterBonus -
+        shipCapacityUsed -
+        getFighterCount(unitCounts);
+
+    const unsupportedFighters =
+        isFighterUpgraded && fighterCapacityRemaining < 0
+            ? Math.abs(fighterCapacityRemaining)
+            : 0;
+
+    const inAdvancedMode =
+        budgetFilters.has(BudgetFilter.FLEET_SUPPLY) ||
+        budgetFilters.has(BudgetFilter.SHIP_CAPACITY);
+
+    const numColumns = 4;
```

Comment:
There's some unfortunate coupling going on here that registers to me as a code smell.

* We configure a `numColumns` here, but it can't really be changed easily. It's actually secretly related/bound to the size of a `FixedSizeRow`, which is a type defined in a whole different file.

* Each instance of a `BudgetRow` looks like it can independently specify columns to hide, but if the `BudgetRows` hide different numbers of columns relative to one another, then I think the Grid will not look as expected.

* The Grid looks like it can define how many columns it has - but it makes no reference to the actual number of columns in the contained rows.

To be clear: none of this looks like it is buggy currently - I think that this works as written. But it seems like it contains a lot of potential for error in the future.

Reply:
So here are some 💭 on potential refactors:

1. Let's not define a `numColumns` at all. Instead, let `numColumns` be inferred by the existing type of `FixedSizeRow` (which is a great idea, btw. Love the type!). That way a `Grid` of `BudgetRows` always has a `FixedSizeRow.size` number of columns. Put a 📌 in hiding columns for now...

2. Even with this change... we could still upset our Grid by adding a child that is NOT a `BudgetRow`. So I would recommend we enforce this constraint at the type level by making this Grid its own component which accepts, as a parameter, a list of `BudgetRow` components. This dovetails nicely with the fact that we want to show/hide rows based on some logic. Now we can just use that logic to construct the `BudgetRow` list correctly, then pass that to the `Grid` which shows everything we pass it.

3. With `Grid` being its own component, `hiddenColumns` can now be given as a prop. This allows the logic for which columns to hide to live outside of the Grid, (at the `BudgetCard` level), it allows for computation of the Grid columns (`FixedSizeRow.size - hiddenColumns.size`), and it ensures that all `BudgetRow` components have the same hidden columns.

4. Admittedly, the Advanced Mode grid cells at the bottom do not fit nicely into this model. It is possible my model here is too weak, but it's also possible these don't really belong in the Grid at all. I notice that both these pieces of information (space dock fighter capacity, and Fighter II research) are kind of "metadata" about the current state of the faction itself, rather than part of a computed budget for the current build. Perhaps they should be handled separately? Even if just placed below the Grid. Thoughts?

Reply:
This is definitely not The One True Way to do this - merely my thoughts. Chew it over and see what you think, feel free to make counterproposals or just stick with what you've got. I do think this is kinda tricky and after all, what you have written does seem to work! So in a way, no changes are strictly necessary :P

Also happy to chat directly about ideas.

Reply:
Your thought process on this is so dang clear and organized. You dialed in on the exact thing I was struggling with the most, I was going in CIRCLES on how to make the types yell at us when the number of columns in the grid got messed up! I'm so happy, you always have a solution for a problem I've given up as a lost cause. 

Diving into it and will report back! 🫡

Reply:
Reporting back with semi-failure, ugh! I couldn't figure out how to do one of these two options:
  - (A) extract the length of `FixedSizeRow` into a variable which I could use in computing `numColumns`

  - (B) the reverse, i.e. pass a variable `numColumns` as a type parameter into `FixedSizeRow` that would set the number of array elements in the type (although while exploring this, I learned how to make a recursive conditional type, which was rad).

For (A), I might have missed an easy way to do this, because I was struggling with how to translate my question into search terms. Do you know how to do this?

I ruled out (B) because eventually the internet explained to me that a variable can't be passed as a type parameter this way, because types are evaluated at compile time while variables are evaluated at runtime.

In case (A) doesn't pan out, I pushed an alternative (worse) solution in [16aa8a9](https://github.com/sirrus233/twilight-imperium-helper/pull/15/commits/16aa8a9a696aedcd910e5264833cb124ab444975), which at least fixes some of the problems you pointed out:

- Removes the potential for a disconnect between `numColumns` and `FixedSizeRow.length`
- Removes the potential for each row to have a different column hidden
- Removes the need to explicitly specify `numColumns` - instead number of columns is inferred from row size
- Throws a runtime error (🤮) if row sizes of a grid don't all match
- Separates the two "meta" inputs at the bottom into their own grid rather than trying to mush them together